### DISCUSSION TITLE
Does phmap benefits from using a custom allocator like tcmalloc?

### DISCUSSION THREAD

**Discussion by ArthoPacini** [greg7mdp/parallel-hashmap#124] [Q&A] [answered]
Is it any beneficial to use some sort of custom allocator like tcmalloc, mimalloc, jemalloc, nedmalloc etc..?

Is there any benchmark comparing these?

Thanks and keep the good work!

**Comment by greg7mdp** [comment]
> Is it any beneficial to use some sort of custom allocator like tcmalloc, mimalloc, jemalloc, nedmalloc etc..?

Probably not for flat_hash_map/set which don't do many allocations.
It may help for node_hash_map/set or btree_map/set, but I don't think it will make a huge difference.
I don't have any benchmark on this.

> Thanks and keep the good work!

Hey, glad you appreciate, and thanks for the kind words!

**Reply by ArthoPacini** [reply]
I am doing some 'amateur' benchmarks (memory usage only, not benchmarking speed) on the flat_hash_map with and without tcmalloc and from what i've seem it just doubles the amount of memory used by the process..
Gonna keep testing some allocators.
Thanks for the fast reply!

**Reply by chausner** [reply]
For my particular application (multi-threaded memory-heavy game tree search), using mimalloc with parallel_flat_hash_set leads to a 43% speedup and 30% reduction in memory usage over the default MSVC STL allocator. I have not tried other custom allocators so far, and I am not sure how far my observations generalize but for me the improvements are massive.

**Comment by greg7mdp** [comment]
Sounds good, thanks @ArthoPacini ! Let us know what you find out.

**Comment by greg7mdp** [comment, marked as answer]
> using mimalloc with parallel_flat_hash_set leads to a 43% speedup and 30% reduction in memory usage over the default MSVC STL allocator

Wow, that's great! I am a bit surprised because the `parallel_flat_hash_set ` itself should not allocate very much, unless you create a large number of them, or use a high number of submaps. Maybe it is the type you insert in the `parallel_flat_hash_set` which allocates? Maybe there are some unnecessary copies causing memory allocations?

**Reply by chausner** [reply]
You're right, in my case I am creating a large number of hash sets and also the element type inside it allocates, as well. So the performance and RAM usage improvements I am seeing with mimalloc are probably not directly related to parallel-hashmap but to my particular use case.
